src/torrent.rs

```diff
@@ -108,86 +115,133 @@ impl Torrent {
-                            self.percent_complete =
-                                self.completed_blocks as f32 / self.total_blocks as f32;
-                        } else {
-                            self.repeated_blocks
-                                .entry((piece.index, b.offset))
-                                .and_modify(|v| *v += 1)
-                                .or_insert(1);
-                        }
-                    }
-                }
-            }
+        let index = self
+            .in_progress_blocks
+            .iter()
+            .position(|block| block.piece_index == piece_index && block.offset == offset)
+            .unwrap_or_else(|| panic!("we should never be trying to fill a piece index and block offset: {:?} that wasn't in the in_progress_blocks field: {:?}", (piece_index, offset), self.in_progress_blocks
+                .iter()
+                .map(|block| {
+                    (block.piece_index, block.offset)
+                })
+            ));
+
+        let b = &mut self.in_progress_blocks[index];
+
+        if b.state != BlockState::Done {
+            b.state = BlockState::Done;
+            b.data = Some(data.to_vec());
+            self.completed_blocks += 1;
+            self.percent_complete = self.completed_blocks as f32 / self.total_blocks as f32;
+            self.completed_pieces[piece_index as usize][block_index as usize] =
+                Some(self.in_progress_blocks.swap_remove(index));
+        } else {
+            self.repeated_blocks
+                .entry((piece_index, offset))
+                .and_modify(|v| *v += 1)
+                .or_insert(1);
         }
     }
 
     pub fn to_file(&self) -> File {
         let file_name = &self.file_name;
         let mut file = File::create(file_name).unwrap();
-        for p in &self.pieces {
-            for b in &p.blocks {
-                let bytes = b.data.as_ref();
-                match bytes {
+        for (piece_index, list_of_filled_blocks) in self.completed_pieces.iter().enumerate() {
+            for (block_index, block) in list_of_filled_blocks.iter().enumerate() {
+                match block {
                     Some(b) => {
-                        file.write_all(b).unwrap();
+                        let bytes = b.data.as_ref();
+                        match bytes {
+                            Some(b) => {
+                                file.write_all(b).unwrap();
+                            }
+                            None => {
+                                println!(
+                                    "missing block {:?} of piece {:?}",
+                                    b.offset, b.piece_index
+                                )
+                            }
+                        }
                     }
                     None => {
-                        println!("missing block {:?} of piece {:?}", b.offset, p.index)
+                        println!(
```

Comment:
This is only possible because list_of_filled_blocks might be None because of completed_pieces having None for unfinished pieces.